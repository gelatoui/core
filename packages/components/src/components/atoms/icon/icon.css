/* ====================== */
/* Default Icon Styles    */
/* ====================== */
:host {
  --glu-icon-size: 24px;
  display: inline-block;
  vertical-align: middle;
}

:host(.glu-icon) {
  color: currentColor;
}
