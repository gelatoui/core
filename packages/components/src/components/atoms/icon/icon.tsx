import { outlineIcons } from './icon-outline'
import { solidIcons } from './icon-solid'

import { Component, h, Host, Prop } from '@stencil/core'

/**
 * A customizable icon component that renders Heroicons
 *
 * @component
 * @tag glu-icon
 * @shadow true
 */
@Component({
  tag: 'glu-icon',
  styleUrl: 'icon.css',
  shadow: true
})
export class GluIcon {
  /**
   * The name of the icon to display
   * @prop {string} name - The kebab-case name of the icon
   * @required
   * @readonly
   * @example "arrow-up"
   * @example "chevron-right"
   */
  @Prop({ reflect: true }) readonly name!: string

  /**
   * The visual variant style of the icon
   * @prop {IconVariant} variant - The styling variant to use
   * @default "outline"
   * @readonly
   */
  @Prop({ reflect: true }) readonly variant: 'solid' | 'outline' = 'outline'

  /**
   * Uniform size of the icon in pixels (takes precedence over width and height)
   * @prop {number} size - When set, overrides both width and height
   * @readonly
   */
  @Prop() readonly size?: number

  /**
   * Custom width of the icon in pixels
   * @prop {number | string} width - Specific width in pixels
   * @default 24
   * @readonly
   */
  @Prop() readonly width: number | string = 24

  /**
   * Custom height of the icon in pixels
   * @prop {number | string} height - Specific height in pixels
   * @default 24
   * @readonly
   */
  @Prop() readonly height: number | string = 24

  /**
   * Custom color for the SVG icon.
   * Para que este cambio tenga efecto, asegÃºrate de que el SVG use `currentColor` en sus atributos.
   * @prop {string} color - Color en cualquier formato CSS (hex, rgb, etc.)
   */
  @Prop() readonly color?: string

  /**
   * Retrieves the SVG content for the configured icon
   * @private
   * @returns {string} SVG markup or empty string if icon not found
   */
  private getIconSvg(): string {
    const icons = this.variant === 'solid' ? solidIcons : outlineIcons

    return icons[this.name] || ''
  }

  private isValidNumber = (value: string | number): boolean => typeof value === 'number' ? Number.isFinite(value) : !isNaN(Number(value.trim()))

  private getSizesValue = ({ width, height }: { width: string | number, height: string | number }):
  { width: string, height: string } => ({
    width: this.isValidNumber(width) ? `${Number(width)}px` : `${width}`,
    height: this.isValidNumber(height) ? `${Number(height)}px` : `${height}`
  })

  /**
   * Renders the icon component inline, allowing CSS color changes
   * @returns {JSX.Element} The rendered icon element
   */
  render() {
    const svgContent = this.getIconSvg()
    const { width, height } = this.getSizesValue({ width: this.size || this.width, height: this.size || this.height })

    if (!svgContent) {
      console.error(`[GluIcon]: Icon "${this.name}" not found.`)

      return null
    }

    // If the content is a data URI, decode it to get the inline SVG
    const svgMarkup = svgContent?.startsWith?.('data:image') ?
      atob(svgContent.split(',')[1]) :
      svgContent

    const style = { width, height, color: this.color || 'inherit' }

    return (
      <Host class="glu-icon">
        <span
          style={style}
          role="img"
          aria-label={this.name}
          innerHTML={svgMarkup}
        />
      </Host>
    )
  }
}
