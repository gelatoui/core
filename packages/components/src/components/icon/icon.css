/* ====================== */
/* Default Icon Styles    */
/* ====================== */
:host {
  display: inline-block;
  vertical-align: middle;
  display: flex;
  justify-content: center;
  align-items: center;
}

span {
  display: flex;
  align-items: center;
  justify-content: center;
}
