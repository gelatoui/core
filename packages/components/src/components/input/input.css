/* ====================== */
/* Input Styles     */
/* ====================== */

:host(.glu-input) {
  gap: var(--Spacing-100, 0.5rem);
  display: flex;
  flex-direction: column;
  width: 100%;
}

.input-wrapper {
  border-radius: var(--Border-Radius-100, 0.5rem);
  border: 1px solid var(--Tokens-Input-Stroke-Default, #CBD1D9);
  background: var(--Tokens-Input-Background-Default, #FFF);
  color: var(--tokens-input-content-default);
  transition: border-color 0.2s, box-shadow 0.2s;
  display: flex;
  flex-direction: row;
  padding: var(--Spacing-150, 0.75rem) var(--Spacing-200, 1rem);
  align-items: center;
  align-self: stretch;
  gap: var(--Spacing-100, 0.5rem);
  flex-shrink: 0;

  color: var(--Tokens-Text-On-Light-Default, #4B515A);

  font-family: var(--Body-Small-Regular-fontFamily, Inter);
  font-size: var(--Body-Small-Regular-fontSize, 0.875rem);
  font-style: normal;
  font-weight: 400;
  line-height: var(--Body-Small-Regular-lineHeight, 1.125rem); /* 128.571% */
  letter-spacing: var(--Body-Small-Regular-letterSpacing, 0rem);

  cursor: text
}

input {
  all: unset;
  width: 100%;
}

input::placeholder {
  color: var(--tokens-input-content-subdue);
}

.prefix-text, .prefix-icon, .suffix-text, .suffix-icon {
  display: flex;
  justify-content: center;
  align-items: center;

  color: var(--Tokens-Text-On-Light-Subdue, #7D8796);

  font-family: var(--Body-Small-Regular-fontFamily, Inter);
  font-size: var(--Body-Small-Regular-fontSize, 0.875rem);
  font-style: normal;
  font-weight: 400;
  line-height: var(--Body-Small-Regular-lineHeight, 1.125rem); /* 128.571% */
  letter-spacing: var(--Body-Small-Regular-letterSpacing, 0rem);
}

:host(.focused) .input-wrapper {
  background-color: var(--tokens-input-background-focus);
  border-color: var(--tokens-input-stroke-focus);
  outline: none;
  box-shadow: 0 0 0 2px var(--tokens-input-stroke-focus);
}

:host(.disabled) .input-wrapper {
  border-radius: var(--Border-Radius-100, 0.5rem);
  border: 1px solid var(--Tokens-Input-Stroke-Disabled, #DCE0E6);
  background: var(--Tokens-Input-Background-Disabled, #F4F5F6);
  cursor: not-allowed;
}

:host(.is-error) .input-wrapper {
  border-color: var(--tokens-input-stroke-error);
  color: var(--tokens-input-content-error);
  background-color: var(--tokens-input-background-error);
  .prefix-text, .prefix-icon, .suffix-text, .suffix-icon {
    color: var(--tokens-input-content-error);
  }
}

:host(.is-error.focused) .input-wrapper {
  box-shadow: 0 0 0 2px var(--tokens-input-stroke-error);
  .prefix-text, .prefix-icon, .suffix-text, .suffix-icon {
    color: var(--tokens-input-stroke-error);
  }
}

input[type="date"]::-webkit-calendar-picker-indicator {
  display: none;
}

input[type="date"]::-moz-calendar-picker-indicator {
  display: none;
}

input[type="date"]::-ms-calendar-picker-indicator {
  display: none;
}

input[type="search"]::-webkit-search-decoration,
input[type="search"]::-webkit-search-cancel-button,
input[type="search"]::-webkit-search-results-button,
input[type="search"]::-webkit-search-results-decoration {
  -webkit-appearance:none;
}

.cursor-pointer {
  cursor: pointer;
}
