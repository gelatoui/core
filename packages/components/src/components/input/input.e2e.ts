/* eslint-disable no-undef */
import { newE2EPage } from '@stencil/core/testing'

describe('glu-input (E2E)', () => {
  it('renders and hydrates', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-input placeholder="Enter text"></glu-input>')

    const element = await page.find('glu-input')

    expect(element).toHaveClass('hydrated')
  })

  it('applies error styling when error prop is true', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-input is-error error="Error occurred"></glu-input>')

    const element = await page.find('glu-input')

    expect(element).toHaveClass('is-error')
  })

  it('renders label and helper text when provided', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-input label="Username" helper-text="Enter your username"></glu-input>')

    const labelEl = await page.find('glu-input >>> glu-label')
    const helperTextEl = await page.find('glu-input >>> glu-helper-text')

    expect(labelEl).not.toBeNull()

    expect(labelEl.textContent).toBe('Username')

    expect(helperTextEl).not.toBeNull()

    expect(helperTextEl.textContent).toContain('Enter your username')
  })

  it('focuses input when host is clicked', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-input placeholder="Focus me"></glu-input>')

    const host = await page.find('glu-input')
    const input = await page.find('glu-input >>> input')

    // Click on the host element.
    await host.click()

    await page.waitForChanges()

    // Evaluate the active element inside the host's shadow root.
    const activeElementTag = await page.evaluate(() => {
      const hostEl = document.querySelector('glu-input')

      return hostEl?.shadowRoot?.activeElement?.tagName
    })

    // The active element should be the input inside our component.
    expect(activeElementTag).toEqual(input.tagName)
  })

  it('updates value on input event', async () => {
    const expectedValue = 'Hello World'
    const page = await newE2EPage()

    await page.setContent('<glu-input></glu-input>')

    const input = await page.find('glu-input >>> input')

    await input.type(expectedValue)

    await page.waitForChanges()

    const inputValue = await input.getProperty('value')

    expect(inputValue).toBe(expectedValue)
  })

  it('toggles password visibility on clicking the toggle icon', async () => {
    const page = await newE2EPage()

    await page.setContent('<glu-input type="password" value="secret"></glu-input>')

    const input = await page.find('glu-input >>> input')

    // Initially, the native input type should be "password".
    expect(await input.getProperty('type')).toBe('password')

    // Click the toggle icon (rendered as a suffix icon).
    const toggleIcon = await page.find('glu-input >>> .suffix-icon')

    await toggleIcon.click()

    await page.waitForChanges()

    // The input type should now be "text".
    expect(await input.getProperty('type')).toBe('text')
  })
})
