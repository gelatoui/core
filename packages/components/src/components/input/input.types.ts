export type InputType =
  | 'button'
  | 'checkbox'
  | 'color'
  | 'date'
  | 'datetime-local'
  | 'email'
  | 'file'
  | 'hidden'
  | 'image'
  | 'month'
  | 'number'
  | 'password'
  | 'radio'
  | 'range'
  | 'reset'
  | 'search'
  | 'submit'
  | 'tel'
  | 'text'
  | 'time'
  | 'url'
  | 'week'

/**
 * Values are converted to strings when emitted which is
 * why we do not have a `number` type here even though the
 * `value` prop accepts a `number` type.
 */
export interface InputChangeEventDetail {
  value?: string
  event?: Event
}

// We recognize that InputInput is not an ideal naming pattern for this type.
export interface InputEventDetail {
  value?: string
  event?: Event
}

export interface InputCustomEvent<T = InputChangeEventDetail> extends CustomEvent {
  detail: T
  target: HTMLInputElement
}
