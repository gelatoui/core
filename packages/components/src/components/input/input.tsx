import { Attributes, inheritAttributes } from '@utils/helpers'

import { InputChangeEventDetail, InputType } from './input.types'

import { Component, Element, Event, EventEmitter, Fragment, h, Host, Prop, State } from '@stencil/core'

/**
 * A composite input component that supports a label, helper text and advanced input features
 * such as prefix/suffix elements, custom input types, and error states.
 *
 * Any extra attributes (except those explicitly defined) set on `<glu-input>` are forwarded
 * to the inner native `<input>` element.
 *
 * @component
 * @tag glu-input
 * @shadow true
 */
@Component({
  tag: 'glu-input',
  styleUrl: 'input.css',
  shadow: true
})
export class GluInput {
  // ---------------------------
  // Field (wrapper) properties
  // ---------------------------
  /**
   * Optional label text displayed above the input field.
   * @prop {string} label - The label text.
   * @readonly
   */
  @Prop() readonly label: string

  /**
   * The error message to display below the input.
   * When non-empty, the component enters an error state and a default error icon is used.
   * @prop {string} error - The error message.
   * @readonly
   */
  @Prop() readonly error: string

  /**
   * Supplemental helper text displayed below the input.
   * @prop {string} helperText - Additional guidance or context.
   * @readonly
   */
  @Prop() readonly helperText: string

  /**
   * The name of the helper icon displayed alongside the helper text.
   * If an error exists, a default error icon is shown instead.
   * @prop {string} helperIcon - The helper icon identifier.
   * @readonly
   */
  @Prop() readonly helperIcon: string

  /**
   * Specifies the visual variant for the helper icon.
   * Allowed values: 'solid' or 'outline'.
   * @prop {'solid' | 'outline'} helperIconVariant - The helper icon style.
   * @default 'outline'
   * @readonly
   */
  @Prop() readonly helperIconVariant: 'solid' | 'outline' = 'outline'

  // ---------------------------
  // Base input properties
  // ---------------------------
  /**
   * Determines the type of the input field.
   * Supported types include 'generic', 'currency', 'phone', 'password', 'date', 'url', 'search', 'number'
   * or any valid InputType. This setting affects both the native input type and UI behavior.
   * @prop {'text' | 'currency' | 'phone' | 'password' | 'date' | 'url' | 'search' | 'number' | InputType} type - The input type.
   * @default 'text'
   * @readonly
   */
  @Prop() readonly type: 'text' | 'currency' | 'phone' | 'password' | 'date' | 'url' | 'search' | 'number' | InputType = 'text'

  /**
   * The name of the suffix icon displayed on the right side of the input.
   * @prop {string} suffixIcon - The suffix icon identifier.
   * @readonly
   */
  @Prop() readonly suffixIcon: string

  /**
   * Defines the visual variant for the suffix icon.
   * Allowed values: 'solid' or 'outline'.
   * @prop {'solid' | 'outline'} suffixIconVariant - The suffix icon style.
   * @default 'outline'
   * @readonly
   */
  @Prop() readonly suffixIconVariant: 'solid' | 'outline' = 'outline'

  /**
   * The name of the prefix icon displayed on the left side of the input.
   * @prop {string} prefixIcon - The prefix icon identifier.
   * @readonly
   */
  @Prop() readonly prefixIcon: string

  /**
   * Defines the visual variant for the prefix icon.
   * Allowed values: 'solid' or 'outline'.
   * @prop {'solid' | 'outline'} prefixIconVariant - The prefix icon style.
   * @default 'outline'
   * @readonly
   */
  @Prop() readonly prefixIconVariant: 'solid' | 'outline' = 'outline'

  /**
   * Optional text displayed as a suffix adjacent to the input field.
   * @prop {string} suffixText - The suffix text content.
   * @readonly
   */
  @Prop() readonly suffixText: string

  /**
   * Optional text displayed as a prefix adjacent to the input field.
   * @prop {string} prefixText - The prefix text content.
   * @readonly
   */
  @Prop() readonly prefixText: string

  /**
   * Renders the input with error styling when true.
   * @prop {boolean} isError - True if the input is in an error state.
   * @default false
   * @readonly
   */
  @Prop() readonly isError: boolean = false

  /**
   * Disables the input field, preventing user interaction.
   * @prop {boolean} disabled - True if the input is disabled.
   * @default false
   * @readonly
   */
  @Prop() readonly disabled: boolean = false

  /**
   * Renders the input field in read-only mode, allowing focus and selection but preventing edits.
   * @prop {boolean} readOnly - True if the input is read-only.
   * @default false
   * @readonly
   */
  @Prop() readonly readOnly: boolean = false

  /**
   * Placeholder text displayed when the input field is empty.
   * @prop {string} placeholder - The placeholder text.
   * @readonly
   */
  @Prop() readonly placeholder: string

  /**
   * The current value of the input field.
   * This property is mutable to support two-way binding.
   * @prop {string} value - The input's value.
   */
  @Prop({ mutable: true, reflect: true }) value

  /**
   * When true and the input type is 'search', displays a clear icon to reset the input.
   * @prop {boolean} showClearIcon - True to display the clear icon.
   * @default false
   * @readonly
   */
  @Prop() readonly showClearIcon: boolean = false

  // ---------------------------
  // Component internal states
  // ---------------------------
  /**
   * Indicates whether the input field currently has focus.
   * @state {boolean} isFocused - True if the input is focused.
   */
  @State() isFocused = false

  /**
   * Toggles the visibility of password text.
   * Only applicable when the input type is 'password'.
   * @state {boolean} showPassword - True if the password is visible.
   */
  @State() showPassword = false

  // ---------------------------
  // References and events
  // ---------------------------
  /**
   * A reference to the host element.
   * @element {HTMLGluInputElement} inputElement - The component's host element.
   */
  // eslint-disable-next-line no-undef
  @Element() inputElement!: HTMLGluInputElement

  /**
   * Emitted when the input value changes.
   * @event {string} gloChanged - The updated value of the input.
   */
  @Event() glChange!: EventEmitter<InputChangeEventDetail>

  /**
   * Emitted when the input element gains focus.
   * @event {string} glFocus
   */
  @Event() glFocus!: EventEmitter<FocusEvent>

  /**
   * Emitted when the input value changes.
   * @event {string} gloChanged
   */
  @Event() glBlur!: EventEmitter<FocusEvent>

  /** Container for attributes inherited from the host element */
  private inheritedAttributes: Attributes = {}

  componentWillLoad() {
    // Inherit attributes from the host element to forward to the inner <input>
    this.inheritedAttributes = { ...inheritAttributes(this.inputElement) }
  }

  // ---------------------------
  // Input event handlers
  // ---------------------------

  private handleInput = (event: Event): void => {
    const input = event.target as HTMLInputElement

    this.value = input.value

    this.glChange.emit({ value: input.value, event })
  }

  private handleFocus = (event: FocusEvent): void => {
    this.isFocused = true

    this.glFocus.emit(event)
  }

  private handleBlur = (event): void => {
    this.isFocused = false

    this.glBlur.emit(event)
  }

  private togglePasswordVisibility = (): void => {
    this.showPassword = !this.showPassword
  }

  private clearInput = (): void => {
    this.value = ''
  }

  // ---------------------------
  // Input type and focus helpers
  // ---------------------------

  private getInputType = (): string => {
    // Determine native input type based on the component's type and state.
    switch (this.type) {
      case 'password':
        return this.showPassword ? 'text' : 'password'

      case 'phone':
        return 'tel'

      case 'number':
        return 'number'

      case 'currency':
        return 'text'

      case 'text':
        return 'text'

      default:
        return this.type
    }
  }

  private focusInput = (): void => {
    // Focus the inner native input element
    const input = this.inputElement.shadowRoot?.querySelector('input')

    input?.focus()
  }

  private focusDateInput = (): void => {
    const input = this.inputElement.shadowRoot?.querySelector('input[type="date"]') as HTMLInputElement

    input?.showPicker?.()

    input?.focus?.()
  }

  // ---------------------------
  // Render helpers for prefix/suffix
  // ---------------------------

  private renderPrefix = () => {
    // Render prefix based on input type or custom prefix props
    switch (this.type) {
      case 'currency':
        return <span class="prefix-text">$</span>

      case 'search':
        return <glu-icon name="magnifying-glass" size={18} class="prefix-icon" />

      case 'date':
        return (
          <glu-icon
            name="calendar"
            size={18}
            class="prefix-icon cursor-pointer"
            variant={this.prefixIconVariant}
            onClick={this.focusDateInput}
          />
        )

      case 'phone':
        return <glu-icon name="phone" size={18} class="prefix-icon" variant={this.prefixIconVariant} />

      case 'url':
        return <span class="prefix-text">https://</span>

      default:
        return (this.prefixIcon || this.prefixText) && (
          <Fragment>
            {this.prefixIcon && (
              <glu-icon name={this.prefixIcon} size={18} class="prefix-icon" variant={this.prefixIconVariant} />
            )}
            {this.prefixText && <span class="prefix-text">{this.prefixText}</span>}
          </Fragment>
        )
    }
  }

  private renderSuffix = () => {
    // Render suffix based on input type or custom suffix props
    switch (this.type) {
      case 'password':
        return (
          <glu-icon
            name={this.showPassword ? 'eye-slash' : 'eye'}
            size={18}
            variant={this.suffixIconVariant}
            class="suffix-icon cursor-pointer"
            onClick={this.togglePasswordVisibility}
          />
        )

      case 'search':
        return this.value && this.showClearIcon && (
          <glu-icon
            name="x-circle"
            size={18}
            class="suffix-icon cursor-pointer"
            variant={this.suffixIconVariant}
            onClick={this.clearInput}
          />
        )

      default:
        return (this.suffixIcon || this.suffixText) && (
          <Fragment>
            {this.suffixIcon && (
              <glu-icon name={this.suffixIcon} size={18} class="suffix-icon" variant={this.suffixIconVariant} />
            )}
            {this.suffixText && <span class="suffix-text">{this.suffixText}</span>}
          </Fragment>
        )
    }
  }

  // ---------------------------
  // Render method
  // ---------------------------

  render() {
    return (
      <Host
        class={{
          'glu-input': true,
          // Apply additional styling if label or helper text is provided
          'glu-input-field': !!this.label || !!this.helperText,
          'is-error': !!this.error || this.isError,
          disabled: this.disabled,
          'read-only': this.readOnly,
          focused: this.isFocused
        }}
      >
        {/* Optional label */}
        {!!this.label && <glu-label>{this.label}</glu-label>}

        <div class="input-wrapper" onClick={this.focusInput}>
          {this.renderPrefix()}
          {/* Inherit additional attributes from host onto the native input */}
          <input
            {...this.inheritedAttributes}
            type={this.getInputType()}
            class="native-input"
            placeholder={this.placeholder}
            value={this.value}
            disabled={this.disabled}
            readOnly={this.readOnly}
            onInput={this.handleInput}
            onFocus={this.handleFocus}
            onBlur={this.handleBlur}
          />
          {this.renderSuffix()}
        </div>

        {/* Optional helper text */}
        {!!(this.error || this.helperText) && (
          <glu-helper-text
            icon={this.helperIcon ?? undefined}
            iconVariant={this.helperIconVariant}
            is-error={!!this.error}
          >
            {this.error || this.helperText}
          </glu-helper-text>
        )}
      </Host>
    )
  }
}
