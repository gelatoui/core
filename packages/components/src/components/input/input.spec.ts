/* eslint-disable no-undef */
import { GluIcon } from '@components/icon/icon'
import { GluInput } from '@components/input/input'

import { newSpecPage } from '@stencil/core/testing'

// TODO: Refactor this test into multiples tests
describe('glu-input', () => {
  it('renders correctly with default props', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input placeholder="Enter text"></glu-input>'
    })

    expect(page.root).toMatchSnapshot()
  })

  it('updates value on input event', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement

    inputEl.value = 'Test value'

    inputEl.dispatchEvent(new Event('input'))

    await page.waitForChanges()

    expect(page.root.value).toBe('Test value')
  })

  it('applies error styling when error prop is true', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input is-error></glu-input>'
    })

    expect(page.root).toHaveClass('is-error')
  })

  it('renders prefix icon when provided', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input prefix-icon="user"></glu-input>'
    })

    const prefixIcon = page.root.shadowRoot.querySelector('.prefix-icon')

    expect(prefixIcon).not.toBeNull()

    expect(prefixIcon.getAttribute('name')).toBe('user')
  })

  it('renders suffix text when provided', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input suffix-text="USD"></glu-input>'
    })

    const suffixText = page.root.shadowRoot.querySelector('.suffix-text')

    expect(suffixText).not.toBeNull()

    expect(suffixText.textContent).toBe('USD')
  })

  it('renders clear icon for search input when showClearIcon is true and value is present', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="search" show-clear-icon value="search term"></glu-input>'
    })

    const clearIcon = page.root.shadowRoot.querySelector('.suffix-icon')

    expect(clearIcon).not.toBeNull()

    expect(clearIcon.getAttribute('name')).toBe('x-circle')
  })

  it('toggles password visibility when password toggle icon is clicked', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="password" value="secret"></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement

    expect(inputEl.getAttribute('type')).toBe('password')

    const toggleIcon = page.root.shadowRoot.querySelector('.suffix-icon') as HTMLElement

    toggleIcon.click()

    await page.waitForChanges()

    expect(inputEl.getAttribute('type')).toBe('text')
  })

  it('triggers focus and blur events correctly', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement
    const spyFocus = jest.fn()
    const spyBlur = jest.fn()

    // Attach event listeners directly to the input element
    inputEl.addEventListener('focus', spyFocus)

    inputEl.addEventListener('blur', spyBlur)

    // Use .focus() and .blur() instead of dispatching an event manually
    inputEl.focus()

    await page.waitForChanges()

    expect(spyFocus).toHaveBeenCalled()

    inputEl.blur()

    await page.waitForChanges()

    expect(spyBlur).toHaveBeenCalled()
  })

  it('disables input when disabled prop is set', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input disabled></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement

    expect(inputEl.hasAttribute('disabled')).toBe(true)
  })

  it('renders placeholder correctly', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input placeholder="Enter text"></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement

    expect(inputEl.getAttribute('placeholder')).toBe('Enter text')
  })

  it('clears the input value when the clear icon is clicked', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="search" show-clear-icon value="search term"></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement
    const clearIcon = page.root.shadowRoot.querySelector('.suffix-icon') as HTMLElement

    expect(inputEl.value).toBe('search term')

    clearIcon.click()

    await page.waitForChanges()

    expect(page.root.value).toBe('')
  })

  it('triggers keydown event when Enter key is pressed', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement
    const spyKeyDown = jest.fn()

    inputEl.addEventListener('keydown', spyKeyDown)

    inputEl.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }))

    await page.waitForChanges()

    expect(spyKeyDown).toHaveBeenCalled()
  })
})

describe('glu-input prefix and suffix', () => {
  it('renders correctly with type "currency"', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="currency" prefix-text="$"></glu-input>'
    })

    const prefixText = page.root.shadowRoot.querySelector('.prefix-text')

    expect(prefixText).not.toBeNull()

    expect(prefixText.textContent).toBe('$')
  })

  it('renders correctly with type "search" and shows clear icon when value is present', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="search" show-clear-icon value="search term"></glu-input>'
    })

    const clearIcon = page.root.shadowRoot.querySelector('.suffix-icon')

    expect(clearIcon).not.toBeNull()

    expect(clearIcon.getAttribute('name')).toBe('x-circle')
  })

  it('toggles visibility for password input when toggle icon is clicked', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="password" value="secret"></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input') as HTMLInputElement

    expect(inputEl.getAttribute('type')).toBe('password')

    const toggleIcon = page.root.shadowRoot.querySelector('.suffix-icon') as HTMLElement

    toggleIcon.click()

    await page.waitForChanges()

    expect(inputEl.getAttribute('type')).toBe('text')

    // Toggle back to password visibility
    toggleIcon.click()

    await page.waitForChanges()

    expect(inputEl.getAttribute('type')).toBe('password')
  })

  it('renders URL input with prefix text "https://"', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="url" prefix-text="https://"></glu-input>'
    })

    const prefixText = page.root.shadowRoot.querySelector('.prefix-text')

    expect(prefixText).not.toBeNull()

    expect(prefixText.textContent).toBe('https://')
  })

  it('renders phone input correctly with appropriate attributes', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="phone"></glu-input>'
    })

    const inputEl = page.root.shadowRoot.querySelector('input[type="tel"]') as HTMLInputElement

    expect(inputEl).not.toBeNull()

    expect(inputEl.getAttribute('type')).toBe('tel')
  })

  it('renders suffix icon for password input', async () => {
    const page = await newSpecPage({
      components: [GluInput, GluIcon],
      html: '<glu-input type="password"></glu-input>'
    })

    const suffixIcon = page.root.shadowRoot.querySelector('.suffix-icon')

    expect(suffixIcon).not.toBeNull()
  })
})
