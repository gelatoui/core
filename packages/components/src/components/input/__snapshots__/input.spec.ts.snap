// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`glu-input renders correctly with default props 1`] = `
<glu-input class="glu-input" placeholder="Enter text">
  <template shadowrootmode="open">
    <div class="input-wrapper">
      <input class="native-input" placeholder="Enter text" type="text">
    </div>
  </template>
</glu-input>
`;
